0000-richer-documents.rst

```diff
@@ -0,0 +1,460 @@
+to as "annotations".::
+
+    data Doc a
+      = Empty
+        -- empty
+      | ...
+      | Pure a -- new! embed an annotation of type 'a'
+
+Such a document model comes with useful instances for standard
+classes, among which ``Functor``, ``Aplicative``, ``Monad``, ``Foldable``
+and ``Traversable``, giving us standard names for functions that
+transform annotations into other annotations or into whole new sub-documents.
+
+``SDoc`` would similarly be updated to take a type parameter for the
+annotation type:::
+
+    newtype SDoc a = SDoc { runSDoc :: SDocContext -> Doc a }
+
+and would come with similar instances as ``Doc`` (since with this definition,
+``SDoc`` is isomorphic to ``ReaderT SDocContext Doc``), except for ``Foldable``
+and ``Traversable``. The main addition to the public API of the ``Outputable``
+module would be a function for constructing a document from an annotation:::
+
+    embed :: a -> SDoc a
+    embed = pure
+
+We could then create an annotation type for error documents, so as to
+embed bits of expressions, types and more, in their AST form, right into
+the documents, even supporting embedding ASTs from all 3 GHC passes. We could
+for example "categorise" error annotations according to the type of entity
+(expression, type, ...) and the AST phase (parser, renamer, typechecker):::
+
+    data PassWrapper f
+      = WrapperPs (f GhcPs) -- parser AST
+      | WrapperRn (f GhcRn) -- renamer AST
+      | WrapperTc (f GhcTc) -- typechecker AST
+
+    data ErrAnnotation
+      = ErrHsExpr (PassWrapper HsExpr)
+        -- ^ expression annotation, from any of the 3 phases
+      | ...
+
+Or alternatively build up a large sum type of all possible types of AST chunks
+that the current error infrastructure emits (as text).::
+
+    data ErrAnnotation
+      = ErrTyCon TyCon
+      -- ^ type constructor annotation, used in "Couldn't match expected type"
+      --   error documents and possibly others
+      | ...
+
+The exact definition of this type will be best figured out while implementing
+this proposal, informed by all the error documents that GHC builds.
+
+We could then start emitting ``ErrAnnotation`` values in our documents
+instead of rendering expressions, types, and friends directly.
+GHC's error storage and reporting infrastructure could then be updated to
+emit ``SDoc ErrAnnotation`` values, with the nice side effect that GHC API users
+(such as developers of IDE tooling) would now get to deal with error documents
+with that type. Most users will quite likely then want to use ``>>=`` to process
+those annotations and render them using GHC's default rendition or a custom
+one:::
+
+    (>>=) :: SDoc a -> (a -> SDoc b) -> SDoc b
+
+This ``Monad`` instance lets us substitute all the annotations in a document
+by fresh, new subdocuments that can depend on the annotation values. Of
+particular interest is the case where ``b`` is ``()`` or ``Void``, indicating
```

Comment:
But `()` and `Void` are different. I think we want `Void` here.

Reply:
Yes, `Void` is more accurate, let's only mention that one.